### PR TITLE
[libc++] Refactor the configuration macros to being always defined

### PR DESCRIPTION
**Pull request by philnik777**
This is a follow-up to #89178. This updates the `<__config_site>` macros.

### CODE REVIEW

**Review by ldionne — CHANGES_REQUESTED** [review]

File: libcxx/include/__config_site.in [hunk 1]
```diff
@@ -11,26 +11,27 @@
...
+#cmakedefine01 _LIBCPP_HAS_TERMINAL
+#cmakedefine01 _LIBCPP_HAS_MUSL_LIBC
+#cmakedefine01 _LIBCPP_HAS_THREAD_API_PTHREAD
+#cmakedefine01 _LIBCPP_HAS_THREAD_API_EXTERNAL
+#cmakedefine01 _LIBCPP_HAS_THREAD_API_WIN32
+#define _LIBCPP_HAS_THREAD_API_C11 0 // FIXME: Is this guarding dead code?
 #cmakedefine _LIBCPP_DISABLE_VISIBILITY_ANNOTATIONS
-#cmakedefine _LIBCPP_HAS_NO_VENDOR_AVAILABILITY_ANNOTATIONS
+#cmakedefine01 _LIBCPP_HAS_VENDOR_AVAILABILITY_ANNOTATIONS
 #cmakedefine _LIBCPP_NO_VCRUNTIME
```
Comment (ldionne) [inline libcxx/include/__config_site.in:26]: Why is there a few macros that aren't migrated over to 0-1?

Comment (philnik777) [reply libcxx/include/__config_site.in:26]: I'm not quite sure what `_LIBCPP_NO_VSRUNTIME` is. We don't ever define it in `CMakeLists.txt` and there seem to be barely any uses in the wild: https://sourcegraph.com/search?q=context:global+_LIBCPP_NO_VCRUNTIME+-file:libcxx/CMakeLists.txt+-file:__config+-file:new+-file:exception+-file:features.py+-file:libcxx/include/BUILD.gn+-file:typeinfo+-file:libcxx/test/config.py+-file:UsingLibcxx.rst&patternType=keyword&sm=0. We should probably just remove it, but that's a different patch.

`_LIBCPP_HAS_NO_STD_MODULES` seems to be dead. We should just remove it in a separate patch.

I've missed `_LIBCPP_INSTRUMENTED_WITH_ASAN`.

**Review by ldionne — APPROVED** [review]
Thanks, LGTM w/ CI green.

**Comment by mstorsjo** [comment]
This caused a surprising issue in external code that is checking these libcxx config macros. Is this change slated to be backported to 19.x, or is this strictly for 20.x development going forward?



Since b4130bee6bfd34d8045f02fc9f951bcb5db9d85c (#109525) we have code in compiler-rt/libfuzzer, which tries to check the kind of thread API used for c++ threads, by doing 
```c++
#if defined(_LIBCPP_HAS_THREAD_API_PTHREAD) ||                                 \
    defined(_GLIBCXX_GCC_GTHR_POSIX_H)
```
Admittedly, this is known to not be a supported construct. (CC @Zentrik)


Now after this change, we suddenly end up picking this pthread codepath every time. The way forward of fixing this, is to extend the condition like this:
```c++
#if (defined(_LIBCPP_HAS_THREAD_API_PTHREAD) && _LIBCPP_HAS_THREAD_API_PTHREAD) || \
    defined(_GLIBCXX_GCC_GTHR_POSIX_H)
```

However with the previous form, we'd have
```c++
#define _LIBCPP_HAS_THREAD_API_PTHREAD
```
where the macro expands to nothing (contrary to `-D _LIBCPP_HAS_THREAD_API_PTHREAD` which would be equal to `-D_LIBCPP_HAS_THREAD_API_PTHREAD=1`).

If we try doing that with the new form of the ifdef, we end up with this error:
```
test.cpp:4:79: error: expected value in expression
    4 | #if (defined(_LIBCPP_HAS_THREAD_API_PTHREAD) && _LIBCPP_HAS_THREAD_API_PTHREAD) || \
      |                                                                               ^
1 error generated.
```

Do you see any way of adjusting the `#if` condition so that it works both with the old and new forms of the define? As the old define doesn't expand to anything, it's hard to do any sort of comparison with it.

If not, I'm afraid that we'd need to rip out the libc++ specific part of #109525, or entirely skip setting the thread names in libfuzzer on mingw targets.

**Comment by ldionne** [comment]
As you mentioned, checking libc++ internal macros is not something we support. If we supported that, we couldn't make any modifications to the code base without breaking people.

Perhaps you could do something along those lines instead?

```c++
if constexpr (std::is_same_v<std::thread::native_handle_type, pthread_t>) {
  // assume pthread
}
```

**Comment by mstorsjo** [comment]
> As you mentioned, checking libc++ internal macros is not something we support. If we supported that, we couldn't make any modifications to the code base without breaking people.
> 
> Perhaps you could do something along those lines instead?
> 
> ```c++
> if constexpr (std::is_same_v<std::thread::native_handle_type, pthread_t>) {
>   // assume pthread
> }
> ```

Hmm, but that expression can't be compiled at all, if the type `pthread_t` isn't declared at this point; in regular builds of libcxx on Windows, there are no pthread headers (which is an optional third party library) included here.

**Comment by mstorsjo** [comment]
> As you mentioned, checking libc++ internal macros is not something we support. If we supported that, we couldn't make any modifications to the code base without breaking people.

Yep - I guess this is a case of us thinking "how bad can it be" (https://xkcd.com/292/).

I guess I assumed that if things change/break (e.g. if the macro is renamed) we'd fail by not picking up libcxx built with pthreads (which, TBH, is a very rare configuration - I think this was added to work with libstdc++ builds on top of pthread, which is a very common configuration). I.e. we assumed that if things break, we'd get false negatives, but instead we get false positives for the condition here.

**Comment by ldionne** [comment]
Here's what I would suggest:

```c++
#if __has_include(<pthread.h>)

#include <pthread.h>
inline void maybe_set_thread_name(std::thread& t, std::string const& name) {
  if constexpr (std::is_same_v<std::thread::native_handle_type, ::pthread_t>) {
    (void)pthread_setname_np(t.native_handle(), name.c_str());
  }
}

#else

inline void maybe_set_thread_name(std::thread& t, std::string const& name) {
  // don't even try
}

#endif
```

**Comment by mstorsjo** [comment]
> Here's what I would suggest:
> 
> ```c++
> #if __has_include(<pthread.h>)
> 
> #include <pthread.h>
> inline void maybe_set_thread_name(std::thread& t, std::string const& name) {
>   if constexpr (std::is_same_v<std::thread::native_handle_type, ::pthread_t>) {
>     (void)pthread_setname_np(t.native_handle(), name.c_str());
>   }
> }
> 
> #else
> 
> inline void maybe_set_thread_name(std::thread& t, std::string const& name) {
>   // don't even try
> }
> 
> #endif
> ```

Thanks - that might indeed work. As the other alternative is to set the thread name with win32 threads, we could also check for that data type - assuming that `pthread_t` and `HANDLE` boil down to distinct types (but it can also very well be that both of them are `void*` underneath...).

Although - I'm potentially leaning towards simply skipping this entirely for mingw, as this is used in libfuzzer which isn't even usable in mingw contexts yet (libfuzzer gets compiled but clang doesn't let you use it); no need to bend over backwards for quality of life improvements for such a configuration.

**Comment by ldionne** [comment]
That's your call!

**Comment by mstorsjo** [comment]
FWIW, that setup with `if constexpr` doesn't work, as the call to `pthread_setname_np` is invalid for these data types, even if `if constexpr` would eliminate it:
```
llvm-project/compiler-rt/lib/fuzzer/FuzzerUtilWindows.cpp:249:11: error: no matching function for call to 'pthread_setname_np'
  249 |     (void)pthread_setname_np(thread.native_handle(), name.c_str());
      |           ^~~~~~~~~~~~~~~~~~
/opt/llvm-mingw/x86_64-w64-mingw32/include/pthread.h:323:26: note: candidate function not viable: cannot convert argument of incomplete type 'native_handle_type' (aka 'void *') to 'pthread_t' (aka 'unsigned long long') for 1st argument
  323 | WINPTHREAD_API int       pthread_setname_np(pthread_t thread, const char *name);
      |                          ^                  ~~~~~~~~~~~~~~~~
1 error generated.
```

I posted a PR to simply skip this on mingw targets, in https://github.com/llvm/llvm-project/pull/115167.

**Comment by ldionne** [comment]
@mstorsjo I think the problem is that your if condition isn't dependent. I think something like this would fix it:

```c++
template <class Thread>
void maybe_set_thread_name(Thread& t, std::string const& name) {
  if constexpr (std::is_same_v<typename Thread::native_handle_type, ::pthread_t>) {
    (void)pthread_setname_np(t.native_handle(), name.c_str());
  }
}
```

This might be worth trying out.

**Comment by mstorsjo** [comment]
> @mstorsjo I think the problem is that your if condition isn't dependent. I think something like this would fix it:
> 
> ```c++
> template <class Thread>
> void maybe_set_thread_name(Thread& t, std::string const& name) {
>   if constexpr (std::is_same_v<typename Thread::native_handle_type, ::pthread_t>) {
>     (void)pthread_setname_np(t.native_handle(), name.c_str());
>   }
> }
> ```
> 
> This might be worth trying out.

Thanks! With that form, I did manage to seemingly get it to work.

I've posted such a solution for future reference, but I think we'll steer clear of that for practical reasons for now. But if we want to go down that route - we now have a proper solution. Thanks!

**Pull request merged**